crates/accelerate/src/quantum_circuit/circuit_data.rs

```diff
@@ -0,0 +1,657 @@
+    /// The intern context used to intern instruction bits.
+    intern_context: InternContext,
+    /// The qubits registered (e.g. through :meth:`~.CircuitData.add_qubit`).
+    qubits_native: Vec<PyObject>,
+    /// The clbits registered (e.g. through :meth:`~.CircuitData.add_clbit`).
+    clbits_native: Vec<PyObject>,
+    /// Map of :class:`.Qubit` instances to their index in
+    /// :attr:`.CircuitData.qubits`.
+    qubit_indices_native: HashMap<BitAsKey, BitType>,
+    /// Map of :class:`.Clbit` instances to their index in
+    /// :attr:`.CircuitData.clbits`.
+    clbit_indices_native: HashMap<BitAsKey, BitType>,
+    /// The qubits registered, cached as a ``list[Qubit]``.
+    qubits: Py<PyList>,
+    /// The clbits registered, cached as a ``list[Clbit]``.
+    clbits: Py<PyList>,
+}
+
+/// A private enumeration type used to extract arguments to pymethods
+/// that may be either an index or a slice.
+#[derive(FromPyObject)]
+pub enum SliceOrInt<'a> {
+    Slice(&'a PySlice),
+    Int(isize),
+}
+
+#[pymethods]
+impl CircuitData {
+    #[new]
+    #[pyo3(signature = (qubits=None, clbits=None, data=None, reserve=0))]
+    pub fn new(
+        py: Python<'_>,
+        qubits: Option<&PyAny>,
+        clbits: Option<&PyAny>,
+        data: Option<&PyAny>,
+        reserve: usize,
+    ) -> PyResult<Self> {
+        let mut self_ = CircuitData {
+            data: Vec::new(),
+            intern_context: InternContext::new(),
+            qubits_native: Vec::new(),
+            clbits_native: Vec::new(),
+            qubit_indices_native: HashMap::new(),
+            clbit_indices_native: HashMap::new(),
+            qubits: PyList::empty(py).into_py(py),
+            clbits: PyList::empty(py).into_py(py),
+        };
+        if let Some(qubits) = qubits {
+            for bit in qubits.iter()? {
+                self_.add_qubit(py, bit?)?;
+            }
+        }
+        if let Some(clbits) = clbits {
+            for bit in clbits.iter()? {
+                self_.add_clbit(py, bit?)?;
+            }
+        }
+        if let Some(data) = data {
+            self_.reserve(py, reserve);
+            self_.extend(py, data)?;
+        }
+        Ok(self_)
+    }
+
+    pub fn __reduce__(self_: &PyCell<CircuitData>, py: Python<'_>) -> PyResult<PyObject> {
+        let ty: &PyType = self_.get_type();
+        let args = {
+            let self_ = self_.borrow();
+            (
+                self_.qubits.clone_ref(py),
+                self_.clbits.clone_ref(py),
+                None::<()>,
+                self_.data.len(),
+            )
+        };
+        Ok((ty, args, None::<()>, self_.iter()?).into_py(py))
+    }
+
+    /// Returns the current sequence of registered :class:`.Qubit`
+    /// instances as a list.
+    ///
+    /// .. note::
+    ///
+    ///     This list is not kept in sync with the container.
+    ///
+    /// Returns:
+    ///     list(:class:`.Qubit`): The current sequence of registered qubits.
```

Comment:
> This list is not kept in sync with the container.

I think this is the right choice for sure.  The fact that it was a direct view onto an internal mutable data structure before was really kind of problematic, since mutating it would make it inconsistent with `_qubit_indices`.